docs/investigations/graphql-partial-updates/graphql-partial-updates.md

```diff
@@ -0,0 +1,156 @@
+# GraphQL Partial Updates
+
+## Overview
+
+This documents describes and compares different approaches to designing partial updates in graphql schema. Most of presented approaches have a link to an example they were based on. 
+
+## Partial update mutation designs
+
+### 1. Simple Patch
+
+#### Schema
+```graphql
+input ApplicationPatchInput {
+    name: String
+    description: String
+    webhooks: [WebhookInput!]
+    healthCheckURL: String
+    apis: [APIDefinitionInput!]
+    eventAPIs: [EventAPIDefinitionInput!]
+    documents: [DocumentInput!]
+}
+
+type Mutation {
+    updateApplication(id: ID!, patch: ApplicationPatchInput!): Application!
+}
+```
+
+#### Description
+In this approach every field has to be optional. If the field is defined, we use the passed value to set the new value. If the field is not defined we don't change the current value.
+
+Because we use the `nil` value to determine whether to change current value or not we lose the ability to set the actual value to `nil`. In the example linked below this limitation doesn't exist because in javascript we can distinguish `null` and `undefined` values.
+
+Example: [link](https://medium.com/workflowgen/graphql-mutations-partial-updates-implementation-bff586bda989)
+
+### 2. Simple Patch (single input)
+
+#### Schema
+```graphql
+input ApplicationUpdateInput {
+    id: ID!,
+    patch: ApplicationPatchInput!
+}
+
+input ApplicationPatchInput {
+    name: String
+    description: String
+    webhooks: [WebhookInput!]
+    healthCheckURL: String
+    apis: [APIDefinitionInput!]
+    eventAPIs: [EventAPIDefinitionInput!]
+    documents: [DocumentInput!]
+}
+
+type Mutation {
+    updateApplication(in: ApplicationUpdateInput!): Application!
+}
+```
+
+#### Description
+This approach is very similar to the previous one, with one difference: there has to be only one, unique input object. In the example linked below it is argued that it makes the client implementation easier.
+
+Patching logic is the same as in previous approach.
+
+Example: [link](https://blog.apollographql.com/designing-graphql-mutations-e09de826ed97)
+
+### 3. Commands & Actions
+
+#### Schema
+```graphql
+input ApplicationUpdateActions {
+    setName: SetApplicationName
+    setDescription: SetApplicationDescription
+    setWebhooks: SetApplicationWebhooks
+    setHealthCheckURL: SetApplicationHealthCheckURL
+    setAPIs: SetApplicationAPIs
+    setEventAPIs: SetApplicationEventAPIs
+    setDocuments: SetApplicationDocuments
+}
+
+input SetApplicationName {
```

Comment:
to check: 
can we define only one type for string input, `SetString`, instead of having `SetApplicationName`, `SetApplicationDescription` ?

Reply:
We would need at least two types, one for optional and one for required string.